### ISSUE TITLE
Running playbooks/deploy_cluster.yml fails on RHEL 7.4

### ISSUE THREAD

**Issue by adelton** [openshift/openshift-ansible#6464]
#### Description

Provide a brief description of your issue here. For example:

On a all-on-one setup on RHEL 7.4, running

```
ansible-playbook -i ansible-inventory openshift-ansible/playbooks/deploy_cluster.yml
```

fails.

##### Version

Please put the following version information in the code block
indicated below.

* Your ansible version per `ansible --version`

```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]
```

If you're operating from a **git clone**:

* The output of `git describe`

```
openshift-ansible-3.9.0-0.8.0
```

##### Steps To Reproduce
1. `ansible-playbook -i ansible-inventory openshift-ansible/playbooks/deploy_cluster.yml`

##### Expected Results
Describe what you expected to happen.

Installation and configuration passing.

##### Observed Results
Describe what is actually happening.

```
TASK [openshift_clock : Determine if chrony is installed] **********************
changed: [master.example.com]

TASK [openshift_clock : Install ntp package] ***********************************
skipping: [master.example.com]

TASK [openshift_clock : Start and enable ntpd/chronyd] *************************
changed: [master.example.com]

TASK [etcd : Set hostname and ip facts] ****************************************
ok: [master.example.com]

TASK [etcd : Add iptables allow rules] *****************************************
failed: [master.example.com] (item={u'port': u'2379/tcp', u'service': u'etcd'}) => {"changed": false, "failed": true, "item": {"port": "2379/tcp", "service": "etcd"}, "module_stderr": "Shared connection to master.example.com closed.\r\n", "module_stdout": "iptables: Bad rule (does a matching rule exist in that chain?).\r\niptables: No chain/target/match by that name.\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_Pm1V6n/ansible_module_os_firewall_manage_iptables.py\", line 283, in <module>\r\n    main()\r\n  File \"/tmp/ansible_Pm1V6n/ansible_module_os_firewall_manage_iptables.py\", line 266, in main\r\n    iptables_manager.add_rule(port, protocol)\r\n  File \"/tmp/ansible_Pm1V6n/ansible_module_os_firewall_manage_iptables.py\", line 88, in add_rule\r\n    self.verify_chain()\r\n  File \"/tmp/ansible_Pm1V6n/ansible_module_os_firewall_manage_iptables.py\", line 81, in verify_chain\r\n    self.create_chain()\r\n  File \"/tmp/ansible_Pm1V6n/ansible_module_os_firewall_manage_iptables.py\", line 207, in create_chain\r\n    self.save()\r\n  File \"/tmp/ansible_Pm1V6n/ansible_module_os_firewall_manage_iptables.py\", line 73, in save\r\n    self.output.append(subprocess.check_output(self.save_cmd, stderr=subprocess.STDOUT))\r\n  File \"/usr/lib64/python2.7/subprocess.py\", line 568, in check_output\r\n    process = Popen(stdout=PIPE, *popenargs, **kwargs)\r\n  File \"/usr/lib64/python2.7/subprocess.py\", line 711, in __init__\r\n    errread, errwrite)\r\n  File \"/usr/lib64/python2.7/subprocess.py\", line 1327, in _execute_child\r\n    raise child_exception\r\nOSError: [Errno 2] No such file or directory\r\n", "msg": "MODULE FAILURE", "rc": 0}
failed: [master.example.com] (item={u'port': u'2380/tcp', u'service': u'etcd peering'}) => {"changed": false, "failed": true, "item": {"port": "2380/tcp", "service": "etcd peering"}, "module_stderr": "Shared connection to master.example.com closed.\r\n", "module_stdout": "iptables: Bad rule (does a matching rule exist in that chain?).\r\niptables: No chain/target/match by that name.\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_tsZAk4/ansible_module_os_firewall_manage_iptables.py\", line 283, in <module>\r\n    main()\r\n  File \"/tmp/ansible_tsZAk4/ansible_module_os_firewall_manage_iptables.py\", line 266, in main\r\n    iptables_manager.add_rule(port, protocol)\r\n  File \"/tmp/ansible_tsZAk4/ansible_module_os_firewall_manage_iptables.py\", line 88, in add_rule\r\n    self.verify_chain()\r\n  File \"/tmp/ansible_tsZAk4/ansible_module_os_firewall_manage_iptables.py\", line 83, in verify_chain\r\n    self.create_jump()\r\n  File \"/tmp/ansible_tsZAk4/ansible_module_os_firewall_manage_iptables.py\", line 178, in create_jump\r\n    self.save()\r\n  File \"/tmp/ansible_tsZAk4/ansible_module_os_firewall_manage_iptables.py\", line 73, in save\r\n    self.output.append(subprocess.check_output(self.save_cmd, stderr=subprocess.STDOUT))\r\n  File \"/usr/lib64/python2.7/subprocess.py\", line 568, in check_output\r\n    process = Popen(stdout=PIPE, *popenargs, **kwargs)\r\n  File \"/usr/lib64/python2.7/subprocess.py\", line 711, in __init__\r\n    errread, errwrite)\r\n  File \"/usr/lib64/python2.7/subprocess.py\", line 1327, in _execute_child\r\n    raise child_exception\r\nOSError: [Errno 2] No such file or directory\r\nChain OS_FIREWALL_ALLOW (0 references)\r\ntarget     prot opt source               destination         \r\n", "msg": "MODULE FAILURE", "rc": 0}

NO MORE HOSTS LEFT *************************************************************
	to retry, use: --limit @/mnt/tests/OpenShift/Origin/install/dist/openshift-ansible/playbooks/deploy_cluster.retry

PLAY RECAP *********************************************************************
master.example.com : ok=114  changed=41   unreachable=0    failed=1   
localhost                  : ok=11   changed=0    unreachable=0    failed=0   


INSTALLER STATUS ***************************************************************
Initialization             : Complete (0:02:49)
Health Check               : Complete (0:05:06)
etcd Install               : In Progress (0:04:44)
	This phase can be restarted by running: playbooks/openshift-etcd/config.yml



Failure summary:


  1. Hosts:    master.example.com
     Play:     Configure etcd
     Task:     Add iptables allow rules
     Message:  All items completed
```

For long output or logs, consider using a [gist](https://gist.github.com/)

##### Additional Information

Provide any additional information which may help us diagnose the
issue.

* This is on RHEL 7.4 and RHEL 7.5 nightly.
* The inventory file is

```
[OSEv3:children]
masters
nodes

[OSEv3:vars]
ansible_ssh_user=root
deployment_type=origin
openshift_disable_check=memory_availability,docker_storage,disk_availability,package_version
openshift_repos_enable_testing=true

osm_cluster_network_cidr=10.128.0.0/14
openshift_portal_net=172.30.0.0/16
osm_host_subnet_length=9

openshift_router_selector='node-router=true'
openshift_registry_selector='node-registry=true'

openshift_master_identity_providers=[{'name': 'allow_all', 'login': 'true', 'challenge': 'true', 'kind': 'AllowAllPasswordIdentityProvider'}]

[masters]
master.example.com

[etcd]
master.example.com

[nodes]
master.example.com openshift_schedulable=true openshift_node_labels="{ 'node-router': 'true', 'node-registry': 'true', 'node-app': 'true' }"
```

The failures started on 2017-12-09, jobs run until 2017-12-08 were passing.

**Comment by adelton** [comment]
Bisecting the commits in master, the last commit where this part of playbook passes is 0beab180d22f28f3f529ca0252f23bb92a7b69a7. So it looks like the regression comes from 8b484281b6052b82869da8cc691e427ec9e1a0fc, https://github.com/openshift/openshift-ansible/pull/6316.

**Comment by michaelgugino** [comment]
@adelton playbooks/prerequisites.yml must now be run before deploy_cluster on new clusters.

**Comment by adelton** [comment]
Thanks, will try that.

It however seems a bit unfortunate -- why can't everything be setup with one step?

**Comment by giuseppe** [comment]
@adelton does running `playbooks/prerequisites.yml ` solve the issue or is it still something that should be investigated in #6316?

**Comment by michaelgugino** [comment]
@adelton Currently, our documentation specifies that a lot of things need to be setup manually before running openshift-ansible, docker (or other container runtime) being one of them.

We currently have a chicken and egg problem when it comes to docker.  We need docker to be installed and configured for health checks, version detection, and image availability to work.  We need all of these steps to happen before we can install the cluster.  In order to maintain the simplest logic, we need to pull docker out into it's own step.

We'll be adding more things into prerequisites; things that need to be installed only once (like iptables).  It doesn't make sense to verify those things are install and configured during every run of the various plays we provide.  The primary reason being that on very large clusters (200+ nodes), every task contributes significantly to the overall execution time.

The trade off is one additional playbook to call, but large time savings for larger clusters.

**Comment by adelton** [comment]
Running `playbooks/prerequisites.yml` fails with

```
TASK [container_runtime : Place additional/blocked/insecure registries in /etc/containers/registries.conf] ***
skipping: [master.example.com]

TASK [container_runtime : Set Proxy Settings] **********************************
skipping: [master.example.com] => (item={u'reg_conf_var': u'HTTP_PROXY', u'reg_fact_val': u''}) 
skipping: [master.example.com] => (item={u'reg_conf_var': u'HTTPS_PROXY', u'reg_fact_val': u''}) 
skipping: [master.example.com] => (item={u'reg_conf_var': u'NO_PROXY', u'reg_fact_val': u''}) 

TASK [container_runtime : Set various Docker options] **************************
fatal: [master.example.com]: FAILED! => {"failed": true, "msg": "template error while templating string: no filter named 'oo_prepend_strings_in_list'. String: OPTIONS='{% if ansible_selinux.status | default(None) == 'enabled' and openshift_docker_selinux_enabled | default(true) | bool %} --selinux-enabled {% endif %} {% if openshift_docker_log_driver | bool %} --log-driver {{ openshift_docker_log_driver }}{% endif %} {% if l2_docker_log_options != [] %} {{ l2_docker_log_options |  oo_split() | oo_prepend_strings_in_list('--log-opt ') | join(' ')}}{% endif %} {% if openshift_docker_hosted_registry_insecure and (openshift_docker_hosted_registry_network | bool) %} --insecure-registry={{ openshift_docker_hosted_registry_network }} {% endif %} {% if docker_options is defined %} {{ docker_options }}{% endif %} {% if openshift_docker_options %} {{ openshift_docker_options }}{% endif %} {% if openshift_docker_disable_push_dockerhub %} --confirm-def-push={{ openshift_docker_disable_push_dockerhub | bool }}{% endif %} --signature-verification={{ openshift_docker_signature_verification | bool }}'"}

RUNNING HANDLER [container_runtime : restart container runtime] ****************
	to retry, use: --limit @/mnt/tests/OpenShift/Origin/install/dist/openshift-ansible/playbooks/prerequisites.retry

PLAY RECAP *********************************************************************
master.example.com : ok=47   changed=10   unreachable=0    failed=1   
localhost                  : ok=11   changed=0    unreachable=0    failed=0   


INSTALLER STATUS ***************************************************************
Initialization             : Complete (0:01:31)

```

but the subsequent `playbooks/deploy_cluster.yml` then passes. Should we try to solve that `prerequisites.yml` failure here or in https://github.com/openshift/openshift-ansible/pull/6316 or should I file separate issue?

**Comment by michaelgugino** [comment]
@adelton Are you using an ansible.cfg file that points to the filter plugins directory?

I am currently refactoring filter plugins into a role that in included via meta-depends by other roles, so we shouldn't see this in the future in any case.

**Comment by adelton** [comment]
Docker has been a separate prerequisite which could be met outside of ansible -- for example it already is there on Atomic hosts. But this new `prerequisites.yml` step was not needed previously, so it seems like and extra complication. Among others, people might need to change the ansible invocation logic and stuff to now process list of two playbooks rather than just one `deploy_cluster.yml`. It seems to go against the logic of "this is how I want my end setup to look like", when you have to manually invoke multiple things in the right order.

I welcome any performance improvements -- I don't have large clusters but I run multiple automated tests and things add up. But I don't really see why checking that iptables are configured appropriately shouldn't be part of the main deployment (and its sanity checking steps).

**Comment by adelton** [comment]
I don't use any nonstandard `ansible.cfg` file, my invocation is

```
ansible-playbook -i ansible-inventory openshift-ansible/playbooks/deploy_cluster.yml
```

**Comment by michaelgugino** [comment]
@adelton I appreciate your feedback.  The information I provided was kind of high-level synopsis.  As you are probably well aware, there are lots of moving pieces to deal with.

Currently, we're working towards 3 main functional areas:
1) prerequisites - things that need to be in place to install openshift.
2) install openshift - the control plane (etcd, masters) and nodes
3) install add-ons - things that run on openshift after it's been deployed.  Registry, router, logging, etc.

After we have these functional areas completed, we may end up re-adding a play that executes all those things together.

We're doing some heavy refactoring at the moment, things are being moved around.  Please keep in mind that the master branch is the primarily development branch, so sometimes things might not go 100% smoothly.

openshift-ansible ships an ansible.cfg in the root directory of the repo.  Our roles and plays currently assume the usage of that config.  One of the items is the filter plugins.  Some logic that has been removed from some areas has been placed into others.

It seems that you need to cd into openshift-ansible or specify that config on the cli/env vars.  However, I'm working on a patch currently that will make that unnecessary for filter plugins.

**Comment by adelton** [comment]
I confirm that setting `ANSIBLE_CONFIG` environment variable to point to `ansible.cfg` in the `openshift-ansible` checkout directory makes `playbooks/prerequisites.yml` pass.

What I see then failing is `playbooks/deploy_cluster.yml` in

```

TASK [template_service_broker : copy] ******************************************
Friday 15 December 2017  05:58:49 -0500 (0:00:00.442)       0:47:02.040 ******* 
changed: [master.example.com] => (item=apiserver-template.yaml)
changed: [master.example.com] => (item=rbac-template.yaml)
changed: [master.example.com] => (item=template-service-broker-registration.yaml)
changed: [master.example.com] => (item=apiserver-config.yaml)

TASK [template_service_broker : yedit] *****************************************
Friday 15 December 2017  05:58:53 -0500 (0:00:03.750)       0:47:05.790 ******* 
ok: [master.example.com]

TASK [template_service_broker : slurp] *****************************************
Friday 15 December 2017  05:58:53 -0500 (0:00:00.460)       0:47:06.251 ******* 
ok: [master.example.com]

TASK [template_service_broker : Apply template file] ***************************
Friday 15 December 2017  05:58:54 -0500 (0:00:00.421)       0:47:06.673 ******* 
changed: [master.example.com]

TASK [template_service_broker : Reconcile with RBAC file] **********************
Friday 15 December 2017  05:58:55 -0500 (0:00:01.214)       0:47:07.887 ******* 
changed: [master.example.com]

TASK [template_service_broker : Verify that TSB is running] ********************
Friday 15 December 2017  05:58:56 -0500 (0:00:01.210)       0:47:09.097 ******* 
FAILED - RETRYING: Verify that TSB is running (120 retries left).
FAILED - RETRYING: Verify that TSB is running (119 retries left).
FAILED - RETRYING: Verify that TSB is running (118 retries left).
FAILED - RETRYING: Verify that TSB is running (117 retries left).
FAILED - RETRYING: Verify that TSB is running (116 retries left).
FAILED - RETRYING: Verify that TSB is running (115 retries left).
FAILED - RETRYING: Verify that TSB is running (114 retries left).
FAILED - RETRYING: Verify that TSB is running (113 retries left).
FAILED - RETRYING: Verify that TSB is running (112 retries left).
FAILED - RETRYING: Verify that TSB is running (111 retries left).
FAILED - RETRYING: Verify that TSB is running (110 retries left).
FAILED - RETRYING: Verify that TSB is running (109 retries left).
FAILED - RETRYING: Verify that TSB is running (108 retries left).
FAILED - RETRYING: Verify that TSB is running (107 retries left).
FAILED - RETRYING: Verify that TSB is running (106 retries left).
FAILED - RETRYING: Verify that TSB is running (105 retries left).
FAILED - RETRYING: Verify that TSB is running (104 retries left).
FAILED - RETRYING: Verify that TSB is running (103 retries left).
FAILED - RETRYING: Verify that TSB is running (102 retries left).
FAILED - RETRYING: Verify that TSB is running (101 retries left).
FAILED - RETRYING: Verify that TSB is running (100 retries left).
FAILED - RETRYING: Verify that TSB is running (99 retries left).
FAILED - RETRYING: Verify that TSB is running (98 retries left).
FAILED - RETRYING: Verify that TSB is running (97 retries left).
FAILED - RETRYING: Verify that TSB is running (96 retries left).
FAILED - RETRYING: Verify that TSB is running (95 retries left).
FAILED - RETRYING: Verify that TSB is running (94 retries left).
FAILED - RETRYING: Verify that TSB is running (93 retries left).
FAILED - RETRYING: Verify that TSB is running (92 retries left).
FAILED - RETRYING: Verify that TSB is running (91 retries left).
FAILED - RETRYING: Verify that TSB is running (90 retries left).
FAILED - RETRYING: Verify that TSB is running (89 retries left).
FAILED - RETRYING: Verify that TSB is running (88 retries left).
FAILED - RETRYING: Verify that TSB is running (87 retries left).
FAILED - RETRYING: Verify that TSB is running (86 retries left).
FAILED - RETRYING: Verify that TSB is running (85 retries left).
FAILED - RETRYING: Verify that TSB is running (84 retries left).
FAILED - RETRYING: Verify that TSB is running (83 retries left).
FAILED - RETRYING: Verify that TSB is running (82 retries left).

MARK-LWD-LOOP -- 2017-12-15 06:00:31 --
FAILED - RETRYING: Verify that TSB is running (81 retries left).
FAILED - RETRYING: Verify that TSB is running (80 retries left).
FAILED - RETRYING: Verify that TSB is running (79 retries left).
FAILED - RETRYING: Verify that TSB is running (78 retries left).
FAILED - RETRYING: Verify that TSB is running (77 retries left).
FAILED - RETRYING: Verify that TSB is running (76 retries left).
FAILED - RETRYING: Verify that TSB is running (75 retries left).
FAILED - RETRYING: Verify that TSB is running (74 retries left).
FAILED - RETRYING: Verify that TSB is running (73 retries left).
FAILED - RETRYING: Verify that TSB is running (72 retries left).
FAILED - RETRYING: Verify that TSB is running (71 retries left).
FAILED - RETRYING: Verify that TSB is running (70 retries left).
FAILED - RETRYING: Verify that TSB is running (69 retries left).
FAILED - RETRYING: Verify that TSB is running (68 retries left).
FAILED - RETRYING: Verify that TSB is running (67 retries left).
FAILED - RETRYING: Verify that TSB is running (66 retries left).
FAILED - RETRYING: Verify that TSB is running (65 retries left).
FAILED - RETRYING: Verify that TSB is running (64 retries left).
FAILED - RETRYING: Verify that TSB is running (63 retries left).
FAILED - RETRYING: Verify that TSB is running (62 retries left).
FAILED - RETRYING: Verify that TSB is running (61 retries left).
FAILED - RETRYING: Verify that TSB is running (60 retries left).
FAILED - RETRYING: Verify that TSB is running (59 retries left).
FAILED - RETRYING: Verify that TSB is running (58 retries left).
FAILED - RETRYING: Verify that TSB is running (57 retries left).
FAILED - RETRYING: Verify that TSB is running (56 retries left).
FAILED - RETRYING: Verify that TSB is running (55 retries left).
FAILED - RETRYING: Verify that TSB is running (54 retries left).
FAILED - RETRYING: Verify that TSB is running (53 retries left).
FAILED - RETRYING: Verify that TSB is running (52 retries left).
FAILED - RETRYING: Verify that TSB is running (51 retries left).
FAILED - RETRYING: Verify that TSB is running (50 retries left).
FAILED - RETRYING: Verify that TSB is running (49 retries left).
FAILED - RETRYING: Verify that TSB is running (48 retries left).
FAILED - RETRYING: Verify that TSB is running (47 retries left).
FAILED - RETRYING: Verify that TSB is running (46 retries left).
FAILED - RETRYING: Verify that TSB is running (45 retries left).
FAILED - RETRYING: Verify that TSB is running (44 retries left).
FAILED - RETRYING: Verify that TSB is running (43 retries left).
FAILED - RETRYING: Verify that TSB is running (42 retries left).
FAILED - RETRYING: Verify that TSB is running (41 retries left).
FAILED - RETRYING: Verify that TSB is running (40 retries left).
FAILED - RETRYING: Verify that TSB is running (39 retries left).
FAILED - RETRYING: Verify that TSB is running (38 retries left).
FAILED - RETRYING: Verify that TSB is running (37 retries left).
FAILED - RETRYING: Verify that TSB is running (36 retries left).
FAILED - RETRYING: Verify that TSB is running (35 retries left).
FAILED - RETRYING: Verify that TSB is running (34 retries left).
FAILED - RETRYING: Verify that TSB is running (33 retries left).
FAILED - RETRYING: Verify that TSB is running (32 retries left).
FAILED - RETRYING: Verify that TSB is running (31 retries left).
FAILED - RETRYING: Verify that TSB is running (30 retries left).
FAILED - RETRYING: Verify that TSB is running (29 retries left).
FAILED - RETRYING: Verify that TSB is running (28 retries left).
FAILED - RETRYING: Verify that TSB is running (27 retries left).
FAILED - RETRYING: Verify that TSB is running (26 retries left).
FAILED - RETRYING: Verify that TSB is running (25 retries left).
FAILED - RETRYING: Verify that TSB is running (24 retries left).
FAILED - RETRYING: Verify that TSB is running (23 retries left).
FAILED - RETRYING: Verify that TSB is running (22 retries left).
FAILED - RETRYING: Verify that TSB is running (21 retries left).
FAILED - RETRYING: Verify that TSB is running (20 retries left).
FAILED - RETRYING: Verify that TSB is running (19 retries left).
FAILED - RETRYING: Verify that TSB is running (18 retries left).
FAILED - RETRYING: Verify that TSB is running (17 retries left).
FAILED - RETRYING: Verify that TSB is running (16 retries left).
FAILED - RETRYING: Verify that TSB is running (15 retries left).
FAILED - RETRYING: Verify that TSB is running (14 retries left).
FAILED - RETRYING: Verify that TSB is running (13 retries left).
FAILED - RETRYING: Verify that TSB is running (12 retries left).
FAILED - RETRYING: Verify that TSB is running (11 retries left).
FAILED - RETRYING: Verify that TSB is running (10 retries left).
FAILED - RETRYING: Verify that TSB is running (9 retries left).
FAILED - RETRYING: Verify that TSB is running (8 retries left).
FAILED - RETRYING: Verify that TSB is running (7 retries left).
FAILED - RETRYING: Verify that TSB is running (6 retries left).
FAILED - RETRYING: Verify that TSB is running (5 retries left).
FAILED - RETRYING: Verify that TSB is running (4 retries left).
FAILED - RETRYING: Verify that TSB is running (3 retries left).
FAILED - RETRYING: Verify that TSB is running (2 retries left).
FAILED - RETRYING: Verify that TSB is running (1 retries left).
fatal: [master.example.com]: FAILED! => {"attempts": 120, "changed": false, "cmd": ["curl", "-k", "https://apiserver.openshift-template-service-broker.svc/healthz"], "delta": "0:00:01.044980", "end": "2017-12-15 06:03:45.417575", "failed": true, "msg": "non-zero return code", "rc": 7, "start": "2017-12-15 06:03:44.372595", "stderr": "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0curl: (7) Failed connect to apiserver.openshift-template-service-broker.svc:443; Connection refused", "stderr_lines": ["  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current", "                                 Dload  Upload   Total   Spent    Left  Speed", "", "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0", "  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0curl: (7) Failed connect to apiserver.openshift-template-service-broker.svc:443; Connection refused"], "stdout": "", "stdout_lines": []}

PLAY RECAP *********************************************************************
localhost                  : ok=12   changed=0    unreachable=0    failed=0   
master.example.com : ok=567  changed=245  unreachable=0    failed=1   


INSTALLER STATUS ***************************************************************
Initialization             : Complete (0:02:27)
Health Check               : Complete (0:05:55)
etcd Install               : Complete (0:06:57)
Master Install             : Complete (0:11:43)
Master Additional Install  : Complete (0:00:46)
Node Install               : Complete (0:12:17)
Hosted Install             : Complete (0:03:55)
Service Catalog Install    : In Progress (0:07:53)
	This phase can be restarted by running: playbooks/openshift-service-catalog/config.yml

Friday 15 December 2017  06:03:45 -0500 (0:04:48.884)       0:51:57.982 ******* 
=============================================================================== 
Run health checks (install) - EL -------------------------------------- 354.79s
template_service_broker : Verify that TSB is running ------------------ 288.88s
etcd : Install etcd --------------------------------------------------- 234.59s
openshift_hosted : Ensure OpenShift pod correctly rolls out (best-effort today) - 130.07s
openshift_cli : Install clients --------------------------------------- 122.04s
Ensure openshift-ansible installer package deps are installed --------- 111.15s
openshift_ca : Install the base package for admin tooling ------------- 106.81s
openshift_node : Install sdn-ovs package ------------------------------ 105.81s
openshift_node : Install Node package --------------------------------- 105.42s
openshift_master : Install Master package ------------------------------ 95.43s
openshift_node : Install Ceph storage plugin dependencies -------------- 95.10s
openshift_node : Install iSCSI storage plugin dependencies ------------- 91.45s
openshift_excluder : Install openshift excluder - yum ------------------ 87.07s
openshift_node : Install dnsmasq --------------------------------------- 85.06s
openshift_node : Install GlusterFS storage plugin dependencies --------- 83.44s
openshift_excluder : Install docker excluder - yum --------------------- 83.42s
openshift_service_catalog : wait for api server to be ready ------------ 40.05s
openshift_hosted : Ensure OpenShift pod correctly rolls out (best-effort today) -- 32.20s
openshift_node : Install conntrack-tools package ----------------------- 28.66s
openshift_node : Install NFS storage plugin dependencies --------------- 28.63s


Failure summary:


  1. Hosts:    master.example.com
     Play:     Service Catalog
     Task:     Verify that TSB is running
     Message:  non-zero return code
```

**Comment by michaelgugino** [comment]
@ewolinetz latest item is tsb failing to deploy on master.  Could you shed any light on this?

**Comment by ewolinetz** [comment]
@adelton Can you `oc describe` the template service broker pod and provide the output here?

**Comment by adelton** [comment]
Not sure which pod you have in mind exactly because I don't see a namespace/pod with `template` in name but I see

```
# oc get pod --all-namespaces
NAMESPACE                          NAME                       READY     STATUS      RESTARTS   AGE
default                            docker-registry-1-2xczg    1/1       Running     0          6h
default                            registry-console-1-v7np5   1/1       Running     0          6h
default                            router-1-8k28g             1/1       Running     0          6h
kube-service-catalog               apiserver-vxfxg            1/1       Running     0          6h
kube-service-catalog               controller-manager-p7lwq   1/1       Running     0          6h
openshift-ansible-service-broker   asb-1-deploy               0/1       Error       0          6h
openshift-ansible-service-broker   asb-etcd-1-deploy          0/1       Error       0          6h
```

and I assume we are after that `asb-1-deploy` which describes as

```
# oc describe -n openshift-ansible-service-broker pod asb-1-deploy
Name:		asb-1-deploy
Namespace:	openshift-ansible-service-broker
Node:		master.example.com/10.16.4.18
Start Time:	Fri, 15 Dec 2017 05:58:36 -0500
Labels:		openshift.io/deployer-pod-for.name=asb-1
Annotations:	openshift.io/deployment.name=asb-1
		openshift.io/scc=restricted
Status:		Failed
IP:		10.128.0.9
Containers:
  deployment:
    Container ID:	docker://c8769e1857775816fd71925b50c2277430a827ee13ba051c4bd115772dd7f965
    Image:		openshift/origin-deployer:v3.7.0
    Image ID:		docker-pullable://docker.io/openshift/origin-deployer@sha256:3d324bce1870047edc418041cefdec88e0a5bbb5b3b9f6fd35b43f14919a656c
    Port:		<none>
    State:		Terminated
      Reason:		Error
      Exit Code:	1
      Started:		Fri, 15 Dec 2017 05:58:42 -0500
      Finished:		Fri, 15 Dec 2017 06:08:44 -0500
    Ready:		False
    Restart Count:	0
    Environment:
      KUBERNETES_MASTER:	https://master.example.com:8443
      OPENSHIFT_MASTER:		https://master.example.com:8443
      BEARER_TOKEN_FILE:	/var/run/secrets/kubernetes.io/serviceaccount/token
      OPENSHIFT_CA_DATA:	-----BEGIN CERTIFICATE-----
MIIC6jCCAdKgAwIBAgIBATANBgkqhkiG9w0BAQsFADAmMSQwIgYDVQQDDBtvcGVu
c2hpZnQtc2lnbmVyQDE1MTMzMzM4OTcwHhcNMTcxMjE1MTAzMTM2WhcNMjIxMjE0
MTAzMTM3WjAmMSQwIgYDVQQDDBtvcGVuc2hpZnQtc2lnbmVyQDE1MTMzMzM4OTcw
ggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQC6/dtlooDCwmXhDVdqNwX+
KSSzVjrrCVuJsjZRWipWKLM84gxmZh87aCblhmEaEGkbe1onB6PCLv0SbTxWZ6Hr
U5SpogFlIiKlJuP1aztOXom7zH9qwP5LSRVbrp9uqMUTFEDtrdxJplHX0uvKAUUv
r0acwbhmzt4gvE98M+jymUO/BQbbOvBh71aQAdy+3pTe1RXCaHe+GTa/HQnmW62/
uEhFDckr0UhENlTidzqK6q8TEvajTr8KQcyHR1gTGtd/Opp/WoB1t5XpHXwf64KF
mGEBcSEs+SFNjj59ESlwATFgzSzmi9gkOvBvVwR+PZ+U+IsGXLP8iuUVUp+9n/uB
AgMBAAGjIzAhMA4GA1UdDwEB/wQEAwICpDAPBgNVHRMBAf8EBTADAQH/MA0GCSqG
SIb3DQEBCwUAA4IBAQAWFgenWZCyAyWzIHpOFlPcaml8mapI4+p1d4B68R3YavDy
NILX1Siyp9RlWplsRMPj7FuwvHVFy4dFwkBUPavbSUGs95UY8XMNLXzK91FnjYDN
ObPBMI6jH2D8YwA0BNiDJfjNzzO1vuAu2EIbw43mII0mX/7rp53WGQhwDt2A8ILm
OEy1dVZxz6WJbA8YiP4GFwT+2C5iSFD9LgZPGrduYQYn1ezSecpDcB9u6WqxRLKH
HktZ4tyz0NcNTkWmmRzoDn6Uk2X83N4tD794K26idYWMPl4pTuzxlNTiyEVNLY91
qlc/aaebo+qdxcdgPPmkYtt0kUwp/RsX2LQG90G/
-----END CERTIFICATE-----

      OPENSHIFT_DEPLOYMENT_NAME:	asb-1
      OPENSHIFT_DEPLOYMENT_NAMESPACE:	openshift-ansible-service-broker
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from deployer-token-xg5sk (ro)
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	False 
  PodScheduled 	True 
Volumes:
  deployer-token-xg5sk:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	deployer-token-xg5sk
    Optional:	false
QoS Class:	BestEffort
Node-Selectors:	<none>
Tolerations:	<none>
Events:		<none>
```

**Comment by ewolinetz** [comment]
There should be a namespace created in the template_service_broker role called `openshift-template-service-broker`.

The pod name should be something like `apiserver`.

Do you get anything back from `oc get ds -n openshift-template-service-broker` ?
Is it possible to rerun just the service-catalog playbook (as described in your error output) but with `-vvv` and provide the output of the task `TASK [template_service_broker : Apply template file]` ?

**Comment by adelton** [comment]
```
# oc get ds -n openshift-template-service-broker
NAME        DESIRED   CURRENT   READY     UP-TO-DATE   AVAILABLE   NODE-SELECTOR   AGE
apiserver   0         0         0         0            0           region=infra    6h
```

I guess the issue might be with that `region=infra` selector because I don't have any nodes with that label. How does one change the selector used by that `ds`?

**Comment by ganhuang** [comment]
The original issue which is being reported is exactly matching https://bugzilla.redhat.com/show_bug.cgi?id=1525357

From the [readme](https://github.com/openshift/openshift-ansible/blob/master/playbooks/README.md), currently we're unable to know what should be run first to deploy the cluster. Hopefully we could add the documentations to clarify the entry playbooks to make it more clear.

**Comment by adelton** [comment]
@ewolinetz, what is the correct way to specify different node selector than that `region=infra`? Assuming I want to run that `apiserver` on nodes with `node-app=true`, how do I do that?

I've tried to use `template_service_broker_selector='node-app=true'` in the inventory file but I get

```
TASK [template_service_broker : Apply template file] ***************************
Tuesday 19 December 2017  06:22:26 -0500 (0:00:00.161)       0:10:43.434 ****** 
fatal: [master.example.com]: FAILED! => {"changed": true, "cmd": "oc process -f \"/tmp/tsb-ansible-LXwoVQ/apiserver-template.yaml\" --param API_SERVER_CONFIG=\"kind: TemplateServiceBrokerConfig\napiVersion: config.templateservicebroker.openshift.io/v1\ntemplateNamespaces:\n- openshift\n\" --param IMAGE=\"docker.io/openshift/origin:latest\" --param NODE_SELECTOR='\"node-app=true\"' | oc apply -f -", "delta": "0:00:00.267278", "end": "2017-12-19 06:22:27.074679", "failed": true, "msg": "non-zero return code", "rc": 1, "start": "2017-12-19 06:22:26.807401", "stderr": "Error from server: cannot restore map from string", "stderr_lines": ["Error from server: cannot restore map from string"], "stdout": "configmap \"apiserver-config\" configured\nserviceaccount \"apiserver\" configured\nservice \"apiserver\" configured\nserviceaccount \"templateservicebroker-client\" configured\nsecret \"templateservicebroker-client\" configured", "stdout_lines": ["configmap \"apiserver-config\" configured", "serviceaccount \"apiserver\" configured", "service \"apiserver\" configured", "serviceaccount \"templateservicebroker-client\" configured", "secret \"templateservicebroker-client\" configured"]}
```

**Comment by milad-soufastai** [comment]
@adelton I've been getting the same sort of errors with the template_service_broker which to me seem like an idempotence issue. I uninstalled my entire cluster using the uninstall.yml file in the adhoc folder, did some docker-storage cleanup, and am rerunning my setup which has worked in the past.

My errors in that step are mainly around the server asking the client to login.

**Comment by milad-soufastai** [comment]
Following up on this: Once I uninstalled the cluster, running the prereq.yml with the ansible.cfg included in the repo and running deploy_cluster.yml works on first run.

**Comment by adelton** [comment]
@milad-soufastai, do you have any node with `region=infra` label?

**Comment by milad-soufastai** [comment]
@adelton I do, I run one master and one node. My master is my infra node.
`osmaster openshift_node_labels="{'region': 'infra','zone': 'default'}" openshift_schedulable=true
osnode1 openshift_node_labels="{'region': 'primary', 'zone': 'east'}"`

**Comment by adelton** [comment]
Well, I don't (as a matter of principle). The output I get in https://github.com/openshift/openshift-ansible/issues/6464#issuecomment-352344496 suggests that the `apiserver` wants to run on `region=infra` node, so that's why it fails for me. You might be hitting some other bug down the road.

**Comment by adelton** [comment]
@ewolinetz, do you have any hints about seeing the TSB running even if the cluster does not have any nodes with `region=infra`? What configuration should be used to force different selector?

**Comment by ewolinetz** [comment]
@adelton apologies, I've been on holiday. To specify that in your inventory you would use the json format `template_service_broker_selector={"node-app": "true"}`

This follows the same pattern as specified in the defaults: https://github.com/openshift/openshift-ansible/blob/master/roles/template_service_broker/defaults/main.yml#L6

**Comment by adelton** [comment]
@ewolinetz, thank you, I confirm that the installation now passes. 

Is there a patch in the works against https://github.com/openshift/openshift-docs to make people aware of `template_service_broker_selector`?

**Comment by adelton** [comment]
To summarize, adding `playbooks/prerequisites.yml` (documentation change https://github.com/openshift/openshift-docs/pull/6668) and `template_service_broker_selector` (due to my setup and tests not using `region=infra`) resolved this issue.

**Comment by CelsoSantos** [comment]
I too faced this issue while installing on Azure based on the MS repo (https://github.com/Microsoft/openshift-container-platform) and the installation failed trying to verify that TSB was running, having to set `template_service_broker_selector={"type": "infra"}` value on the inventory file, otherwise the process would fail.

I noticed that nodes are getting labeled with `type` but not with `region`, thus the reason for failure.

I'm not sure that running `prerequisites.yml` will help with this, but I leave my conclusions above

**Comment by vrutkovs** [comment]
>having to set template_service_broker_selector={"type": "infra"} value on the inventory file, otherwise the process would fail.

Default TSB (and any hosted services) selector is `region: infra`, which is recommended in `hosts.example`. It seems Microsoft/openshift-container-platform uses `type: infra` instead (https://github.com/Microsoft/openshift-container-platform/blob/6244574eeab4e02602c38fe7813f6ffa629eea39/scripts/infraAdd.sh#L21) thus the failures
